src/main/java/org/commcare/formplayer/application/FormController.java

```diff
@@ -207,6 +215,8 @@ public SubmitResponseBean submitForm(@RequestBody SubmitRequestBean submitReques
 
                 // Only delete session immediately after successful submit
                 deleteSession(submitRequestBean.getSessionId());
+                String cacheKey = UserUtils.getFullUserDetail(submitRequestBean.getUsername(), submitRequestBean.getRestoreAs(), submitRequestBean.getDomain());
```

Comment:
I think you would be better off isolating / abstracting these cache management operations in the **RestoreFactory** or in the UserSqlSandbox, which is retrieved from the restore factory

Those dependencies are autowired everywhere that you are currently managing them, and already have access to the key components in a more centralized way, which reduces the risk of drift between how they are established in different contexts.

That would also let you centralize a few of the cache clearing operations, although based on the size of the cache I think it might make more sense to encode DB state in the key rather than clearing it deliberately